### ISSUE TITLE
TS2769: No overload matches this call

### ISSUE THREAD

**Issue by viT-1** [vuejs/vue-class-component#374]
Typescript: 3.7.2
vue-property-decorator: 8.3.0
Vue: 2.6.10

All compiles fine, and I see component in browser which is working as expected, BUT
I forced to use ts-ignore on Vue initialization =(
Just uncomment [this line](https://github.com/viT-1/systemjs-ts-es6-vue/blob/5b812fbeca8c5a8b01e855f430526fcbf7935306/src/VueApp/VueApp.ts#L24),
and start `npm run deploy.dev` and you catch this error:
```
error TS2769: No overload matches this call.
  Overload 1 of 3, '(options?: ThisTypedComponentOptionsWithArrayProps<Vue, { mySelected: null; }, object, object, "$el" | "$options" | "$parent" | "$root" | "$children" | "$refs" | "$slots" | "$scopedSlots" | "$isServer" | "$data" | "$props" | ... 15 more ... | "$createElement"> | undefined): CombinedVueInstance<...>', gave the following error.
    Argument of type '{ el: string; components: { IamHeader: typeof IamHeader; IamSelect: typeof IamSelect; }; data: { mySelected: null; }; }' is not assignable to parameter of type 'ThisTypedComponentOptionsWithArrayProps<Vue, { mySelected: null; }, object, object, "$el" | "$options" | "$parent" | "$root" | "$children" | "$refs" | "$slots" | "$scopedSlots" | "$isServer" | "$data" | "$props" | ... 15 more ... | "$createElement">'.
      Type '{ el: string; components: { IamHeader: typeof IamHeader; IamSelect: typeof IamSelect; }; data: { mySelected: null; }; }' is not assignable to type 'ComponentOptions<Vue, DataDef<{ mySelected: null; }, Record<"$el" | "$options" | "$parent" | "$root" | "$children" | "$refs" | "$slots" | "$scopedSlots" | "$isServer" | "$data" | "$props" | "$ssrContext" | ... 14 more ... | "$createElement", any>, Vue>, object, object, ("$el" | ... 25 more ... | "$createElement")[],...'.
        Types of property 'components' are incompatible.
          Type '{ IamHeader: typeof IamHeader; IamSelect: typeof IamSelect; }' is not assignable to type '{ [key: string]: VueConstructor<Vue> | FunctionalComponentOptions<any, PropsDefinition<any>> | ComponentOptions<never, any, any, any, any, Record<...>> | AsyncComponentPromise<...> | AsyncComponentFactory<...>; }'.
            Property 'IamHeader' is incompatible with index signature.
              Type 'typeof IamHeader' is not assignable to type 'VueConstructor<Vue> | FunctionalComponentOptions<any, PropsDefinition<any>> | ComponentOptions<never, any, any, any, any, Record<...>> | AsyncComponentPromise<...> | AsyncComponentFactory<...>'.
                Type 'typeof IamHeader' is not assignable to type 'VueConstructor<Vue>'.
                  Type 'IamHeader' is not assignable to type 'CombinedVueInstance<Vue, object, object, object, Record<never, any>>'.
                    Type 'IamHeader' is missing the following properties from type 'Vue': $el, $options, $parent, $root, and 23 more.
  Overload 2 of 3, '(options?: ThisTypedComponentOptionsWithRecordProps<Vue, { mySelected: null; }, object, object, { readonly $el: unknown; readonly $options: unknown; readonly $parent: unknown; readonly $root: unknown; readonly $children: unknown; readonly $refs: unknown; readonly $slots: unknown; ... 19 more ...; $createElement: unknown; }> | undefined): CombinedVueInstance<...>', gave the following error.
    Argument of type '{ el: string; components: { IamHeader: typeof IamHeader; IamSelect: typeof IamSelect; }; data: { mySelected: null; }; }' is not assignable to parameter of type 'ThisTypedComponentOptionsWithRecordProps<Vue, { mySelected: null; }, object, object, { readonly $el: unknown; readonly $options: unknown; readonly $parent: unknown; readonly $root: unknown; readonly $children: unknown; readonly $refs: unknown; readonly $slots: unknown; readonly $scopedSlots: unknown; ... 18 more ......'.
      Type '{ el: string; components: { IamHeader: typeof IamHeader; IamSelect: typeof IamSelect; }; data: { mySelected: null; }; }' is not assignable to type 'ComponentOptions<Vue, DataDef<{ mySelected: null; }, { readonly $el: unknown; readonly $options: unknown; readonly $parent: unknown; readonly $root: unknown; readonly $children: unknown; readonly $refs: unknown; readonly $slots: unknown; readonly $scopedSlots: unknown; ... 18 more ...; $createElement: unknown; }, Vu...'.
        Types of property 'components' are incompatible.
          Type '{ IamHeader: typeof IamHeader; IamSelect: typeof IamSelect; }' is not assignable to type '{ [key: string]: VueConstructor<Vue> | FunctionalComponentOptions<any, PropsDefinition<any>> | ComponentOptions<never, any, any, any, any, Record<...>> | AsyncComponentPromise<...> | AsyncComponentFactory<...>; }'.
  Overload 3 of 3, '(options?: ComponentOptions<Vue, DefaultData<Vue>, DefaultMethods<Vue>, DefaultComputed, PropsDefinition<Record<string, any>>, Record<...>> | undefined): CombinedVueInstance<...>', gave the following error.
    Argument of type '{ el: string; components: { IamHeader: typeof IamHeader; IamSelect: typeof IamSelect; }; data: { mySelected: null; }; }' is not assignable to parameter of type 'ComponentOptions<Vue, DefaultData<Vue>, DefaultMethods<Vue>, DefaultComputed, PropsDefinition<Record<string, any>>, Record<...>>'.
      Types of property 'components' are incompatible.
        Type '{ IamHeader: typeof IamHeader; IamSelect: typeof IamSelect; }' is not assignable to type '{ [key: string]: VueConstructor<Vue> | FunctionalComponentOptions<any, PropsDefinition<any>> | ComponentOptions<never, any, any, any, any, Record<...>> | AsyncComponentPromise<...> | AsyncComponentFactory<...>; }'.
          Property 'IamHeader' is incompatible with index signature.
            Type 'typeof IamHeader' is not assignable to type 'VueConstructor<Vue> | FunctionalComponentOptions<any, PropsDefinition<any>> | ComponentOptions<never, any, any, any, any, Record<...>> | AsyncComponentPromise<...> | AsyncComponentFactory<...>'.
              Type 'typeof IamHeader' is not assignable to type 'VueConstructor<Vue>'.
TypeScript: 1 semantic error
TypeScript: emit failed
[10:57:17] 'transpile.dev' errored after 1.56 s
[10:57:17] Error: TypeScript: Compilation failed
    at Output.mightFinish (C:\Projects\systemjs-ts-es6-vue\node_modules\gulp-typescript\release\output.js:131:43)
    at Output.finish (C:\Projects\systemjs-ts-es6-vue\node_modules\gulp-typescript\release\output.js:123:14)
    at ProjectCompiler.inputDone (C:\Projects\systemjs-ts-es6-vue\node_modules\gulp-typescript\release\compiler.js:97:29)
    at CompileStream.end (C:\Projects\systemjs-ts-es6-vue\node_modules\gulp-typescript\release\project.js:125:31)
    at DestroyableTransform.onend (C:\Projects\systemjs-ts-es6-vue\node_modules\readable-stream\lib\_stream_readable.js:577:10)
    at Object.onceWrapper (events.js:286:20)
    at DestroyableTransform.emit (events.js:203:15)
    at DestroyableTransform.EventEmitter.emit (domain.js:466:23)
    at endReadableNT (C:\Projects\systemjs-ts-es6-vue\node_modules\readable-stream\lib\_stream_readable.js:1010:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

`Type 'typeof IamHeader' is not assignable to type 'VueConstructor<Vue>`
[from src](https://github.com/viT-1/systemjs-ts-es6-vue/blob/5b812fbeca8c5a8b01e855f430526fcbf7935306/src/common.blocks/IamHeader/IamHeader.ts#L13)
```
@Component({
	name,
	template,
})
export class IamHeader extends Mixins(BemComponent) {
}
```

[from src](https://github.com/viT-1/systemjs-ts-es6-vue/blob/5b812fbeca8c5a8b01e855f430526fcbf7935306/src/common.blocks/BemComponent/BemComponent.ts#L9)
```
@Component
export class BemComponent extends Vue
	implements IBem, IClassComponent {
}
```

**Comment by viT-1** [comment]
`npm run deploy` is working without ts-ignore & without errors
differences in [tsconfig.dev.json](https://github.com/viT-1/systemjs-ts-es6-vue/blob/5b812fbeca8c5a8b01e855f430526fcbf7935306/src/tsconfig.dev.json) vs [tsconfig.json](https://github.com/viT-1/systemjs-ts-es6-vue/blob/5b812fbeca8c5a8b01e855f430526fcbf7935306/src/tsconfig.json)

**Comment by viT-1** [comment]
`"strictFunctionTypes": false` is not helped to me

**Comment by ktsn** [comment]
Please minimize your reproduction to make sure your problem is actually caused by vue-class-component. e.g. Remove extra build tool / configuration (gulp, multiple tsconfig.json etc.), only include codes which causes this problem.
See how to make reproduction: https://new-issue.vuejs.org/?repo=vuejs/vue#why-repro

**Comment by viT-1** [comment]
@ktsn ok I am working on it

**Comment by decentraliser** [comment]
Hi guys, any update about this issue?

**Comment by viT-1** [comment]
@decentraliser I can't reproduct this problem on another small project.

**Comment by decentraliser** [comment]
it happened to me when I deleted `package-lock.json` and `/node_modules`, then `npm i`.
not sure if I can reproduce it
afterwards I just cloned my repo in a new folder and everything worked

**Comment by viT-1** [comment]
@decentraliser I just tried your way, but deleting `node_modules` folder & `package-lock.json`.

**Comment by PankajChibhrani** [comment]
I am also facing the same issue

**Comment by hutber** [comment]
Also no idea how to get past this issue

**Comment by ktsn** [comment]
Closing until a minimal reproduction is provided. If you are able to provide it which clarify the problem is caused by vue-class-component, please open a new issue.

**Comment by ericksonc** [comment]
Just in case anyone else is getting similar "no overload matches this call" errors, I was having a similar issue when I was trying to import a component defined in vanilla JS.

The main thing is that the rule about needing to use Vue.extend({}) applies to child components as well.

To fix it I just needed to change this:
```

import MyComponent from '...'
@Component({ 
  components: {
    "myComponent": myComponent
  } 
})
```

Into this:

```
import MyComponent from '...'
@Component({ 
  components: {
    "myComponent": Vue.extend(myComponent)
  } 
})
```

**Comment by NikoGrano** [comment]
#### My mistake which caused this:

I had same issue, however, I was able to fix it. Propably nobody else will have totally same retarded mistake as I made, but this might help in debugging.

So, I had code 

```ts
    @Component({
      components: {
        HStockWidgetItemCount,
        HStockWidgetActions,
        HStockTableProducts
      }
    })
    export default class StockOverview extends Vue {
      readonly name: string = 'StockOverview';
    }
```

This was causing `TS2769`. Thanks to @ericksonc I tried [his](https://github.com/vuejs/vue-class-component/issues/374#issuecomment-586558124) way to fix this. So I end up with following code:

```ts
    @Component({
      components: {
        'h-stock-widget-item-count': Vue.extend(HStockWidgetItemCount),
        'h-stock-widget-actions': Vue.extend(HStockWidgetActions),
        'h-stock-table-products': Vue.extend(HStockTableProducts)
      }
    })
    export default class StockOverview extends Vue {
      readonly name: string = 'StockOverview';
    }
```

Well, this did either work. I end up with `ERROR in /.../HStockWidgetActions.vue.ts(29,2)`. So this error was caused totally by faulty module. In my case the fault in this module was missing the extending of Vue.

Meaning by **wrong** code was following 
```ts
    import { Component } from "vue-property-decorator";

    @Component
    export default class HStockWidgetActions {
        public readonly name: string = 'HStockWidgetActions';
    }
```
while it supposed be
```ts
    import { Vue, Component } from "vue-property-decorator";

    @Component
    export default class HStockWidgetActions extends Vue {
        public readonly name: string = 'HStockWidgetActions';
    }
```

After this I just put back the "correct" code and it passed without problems. 
```ts
    @Component({
      components: {
        HStockWidgetItemCount,
        HStockWidgetActions,
        HStockTableProducts
      }
    })
    export default class StockOverview extends Vue {
      readonly name: string = 'StockOverview';
    }
```
#### Possible another solution

I remember also having this issue few weeks before also. When this happend after I switched node from 10 to 13 with `nvm`. Solution to this was just drop the node_modules folder and install again. Problem solved with that.


Just one question rises... Why the heck Javascript/TypeScript packages are so easy to break and why they dont have proper stacktrace to debug.... :expressionless:

**Comment by laurieskelly** [comment]
> #### Possible another solution
> I remember also having this issue few weeks before also. When this happend after I switched node from 10 to 13 with `nvm`. Solution to this was just drop the node_modules folder and install again. Problem solved with that.

This helped me to solve my problem, to anyone finding this googling the error code. After double-checking my `export`s and `extend`s and my spelling of everything, and having the problem persist, refreshing my project packages with `yarn install` resolved this very annoying issue. Thank you @NikoGrano for the footnote!

**Comment by viT-1** [comment]
@ktsn 
[Reproduction project](https://github.com/viT-1/systemjs-ts-es6-vue/blob/95c9323e543cbedddb2604c441f3b50e434b3593/package.json#L56)
[Problem code](https://github.com/viT-1/systemjs-ts-es6-vue/blob/95c9323e543cbedddb2604c441f3b50e434b3593/src/VueApp/VueApp.ts#L33) (as soon as updating vue-property-decorator from v8.4.2 to v9.1.2)

**Comment by lighth7015** [comment]
That isn't a reproduction, though. A reproduction is a 5 file (or less) "hello world" project, so you can eliminate it's not some other issue.

**Comment by agrantham** [comment]
For anyone coming here and none of the above solutions worked for you I wanted to share why I was getting this error and how I fixed it. The errors being shown for me were not readily apparent and did not help narrow down the problem.

I had all working code and class based components working in my NuxtJS project. I added a global mixin computed property and then all these errors started showing up:
![Screen Shot 2022-03-10 at 12 05 56 PM](https://user-images.githubusercontent.com/2923432/157745806-cca6a961-cecb-44fa-80f6-fa389bd410e7.png)

It was very confusing and I didn't know why. I then started traversing down into the components being imported. All the spelling was correct and all components were being property extended.

Then I realized that one of the nested components had a Property name with the same name as the computed property I had in the mixin:
![Screen Shot 2022-03-10 at 12 08 12 PM](https://user-images.githubusercontent.com/2923432/157746195-43731524-79a1-40c6-9e00-2f5e68086c25.png)

Once I removed the property value since it was no longer needed all of the TS errors went away:
![Screen Shot 2022-03-10 at 12 09 17 PM](https://user-images.githubusercontent.com/2923432/157746298-b797d0e9-6582-4628-8659-ea2152d7826b.png)

I hope this was helpful for anyone else coming here by googling that TS error! Good luck!!
